compatibility/compatibility-vec12to20/src/main/java/com/foursoft/harness/compatibility/vec12to20/Vec12XTo20XClassMapper.java

```diff
@@ -0,0 +1,103 @@
+/*-
+ * ========================LICENSE_START=================================
+ * Compatibility VEC 1.2.X To VEC 2.0.X
+ * %%
+ * Copyright (C) 2020 - 2023 4Soft GmbH
+ * %%
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ * =========================LICENSE_END==================================
+ */
+package com.foursoft.harness.compatibility.vec12to20;
+
+import com.foursoft.harness.compatibility.core.HasUnsupportedMethods;
+import com.foursoft.harness.compatibility.core.MethodIdentifier;
+import com.foursoft.harness.compatibility.core.mapping.NameBasedClassMapper;
+import com.foursoft.harness.compatibility.core.util.ClassUtils;
+import com.foursoft.harness.vec.v2x.VecCableLeadThrough;
+
+import java.io.Serial;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Class responsible for mapping VEC 1.2.X classes to VEC 2.0.X <b>and vice versa</b>.
+ */
+public class Vec12XTo20XClassMapper extends NameBasedClassMapper {
+
+    private final Map<Class<?>, Class<?>> classMap;
+    private final UnsupportedVec12XToVec20XMethods ignored;
+
+    /**
+     * Creates a VEC 1.1 to VEC 1.2 Class Mapper.
+     */
+    public Vec12XTo20XClassMapper() {
+        super(Constants.PACKAGE_VEC12X, Constants.PACKAGE_VEC20X);
+
+        classMap = new HashMap<>();
+
+        // VEC 1.2.X -> VEC 2.0.X
+
+        // VEC 2.0.X -> VEC 1.2.X
```

Comment:
I don't get your point. The ClassMapper need dependencies to it's 2 VEC versions for conversion. But there is no module with dependencies to all 3 versions. And this refectoring is not enough to do so in my opinion.

Reply:
Currently, `classMap` is empty since there is no addition to it. Due to that, the `map` method also always executes the else case. In general, it just returns `super.map(clazz)`. So I would either simplify it to that or refactor it so there is just one `VecClassMapper`. For example it could take the "conversion maps" as constructor parameters. Because so fair, `classMap` uses wildcards and `UnsupportedVec12XToVec20XMethods` has a `Map<MethodIdentifier>`.

And I would prefer the refactoring.

Reply:
@DManstrator you are right, there is no addition to Map. If the intention of the `classMap` is to provide a "cache" of the calculated values, then are some problems with that:

1. @TobiasFehlmann Maybe this is a misunderstanding of the `Map.getOrDefault` method. This method returns a default value, if there is no value for the key in the map. However, the calculated value is not placed in the map. It is a bassically a "substitution" method for not returning null. E.g. in a `xy->String` map for translations the default might be `"Not Found"`.
2. If using `getOrDefault`, the default value should be a constant. Since the parameter is passed as a value the `super.map(clazz)` method is called always, regardless if its value is required or not. So, even if `getOrDefault` would be a valid caching approach, it would be destroyed be the call to `super.map(clazz)` because this is executed every time line 76 is executed.
3. When you want to use a map as cache you do the following:
```java 
   public Class<?> map(final Class<?> clazz) {
        final Class<?> aClass = classMap.get(clazz);
        return aClass != null ? aClass : classMap.computeIfAbsent(ClassUtils.getNonProxyClass(clazz), super::map);
    }
```

Reply:
The `classMap` has the use to extend the mapping of classes from one package to another package.

In most cases, a class (e.g. `VecPartOccurrence`) is present in both packages. So the basic code is "replace the source package with the target package and try to find the class".

Example: `com.foursoft.harness.vec.v113.VecPartOccurrence` -> `com.foursoft.harness.vec.v12x.VecPartOccurrence`. Exists? Yes.

 However there are classes where this is not the case, e.g. VEC 1.1.X having `VecZone` and VEC 1.2.X having `VecTopologyZone` instead. If the default `map` method would be invoked, an exception would be thrown since the class can't be found in the other package.

I added the `getOrDefault` part when I recognized that `clazz` was sometimes a proxy class and sometimes not. So sometimes the class could be found in the `classMap`, sometimes not. So I added that as a fail safe.

So what is happening is basically this:

1. Is `clazz` present in `classMap`? (Checks if the given class needs a different class mapping.)
2. If yes, then simply return the mapped class provided by `classMap`.
3. If no, check: Is the non-proxy class of `clazz` is present in `classMap`? (Failsafe check.)
4. If yes, then simply return the mapped class provided by `classMap`.
5. If no, execute the super implementation.

Since this is confusing, I would add comments to the code or making it more verbose to make it more clear.

Reply:
@DManstrator the problem is not that it is confusing what happens, that is quite clear. However, the point is, that it is sloppy programming style, because (as mentioned above) `super.map(clazz)` is executed always when line 76 is entered, no matter if the result is required or not. This is wasted execution time and can be, depending on the scenario quite expensive. In this this case, I don't think it really matters. I'll post a suggustion for a much clean solution.

Reply:
The refactoring has still to be done, will happen in a later PR.